JAVA-2928: Detect if transactions are not supported by the server

jstewart148 opened the pull request:
If a transaction is started, check the max wire version to detect if transactions are supported. For the sync code, the wire version is checked in `ClientSessionImpl#startTransaction`. For the async code, the wire version is checked in `OperationExecutorImpl#execute` by adding callbacks.
Evergreen patch: https://evergreen.mongodb.com/version/5c47eefc3e8e8678e41c6672

jstewart148 committed:
JAVA-2928: add test

jstewart148 committed:
JAVA-2928: detect if transactions are not supported by the server

jstewart148 committed:
JAVA-2928: add transaction failure test

jstewart148 committed:
JAVA-2928: added debug output to exception

jstewart148 committed:
JAVA-2928: add async code for transaction detection

jstewart148 committed:
JAVA-2928: async changes for txn detection

jstewart148 committed:
JAVA-2928: update maxVersion code

jstewart148 committed:
JAVA-2928: adjust sync code to detect transactions

jyemin commented on driver-sync/src/main/com/mongodb/client/internal/ClientSessionImpl.java:
```diff
@@ -102,6 +108,21 @@ public void startTransaction(final TransactionOptions transactionOptions) {
     }
 
+    private void checkServerVersionForTransactionSupport() {
+        ClusterDescription clusterDescription = delegate.getCluster().getCurrentDescription();
```
Do you mean to be using the non-blocking call here?  I'm not convinced it's the right thing to do.  What would happen if this is the first thing the application does after creating the MongoClient?

jstewart148 replied:
I did not intentionally mean to use a non-blocking call here. I've changed the call.

jyemin commented on driver-sync/src/main/com/mongodb/client/internal/ClientSessionImpl.java:
```diff
@@ -102,6 +108,21 @@ public void startTransaction(final TransactionOptions transactionOptions) {
+                break;
+            }
+        }
+        if (wireVersion < 7
```
You can use the constants in com.mongodb.internal.operation.ServerVersionHelper instead.

jstewart148 replied:
Done

jyemin commented on driver-sync/src/main/com/mongodb/client/internal/ClientSessionImpl.java:
```diff
@@ -102,6 +108,21 @@ public void startTransaction(final TransactionOptions transactionOptions) {
+        int wireVersion = 0;
+        for (ServerDescription serverDescription : clusterDescription.getServerDescriptions()) {
+            if (serverDescription.getType() != ServerType.UNKNOWN) {
+                wireVersion = serverDescription.getMaxWireVersion();
```
Should we check all the servers instead of stopping at the first and use the minimum wire version?  The way it is now there is randomness depending on which server description we happen to encounter first.

jstewart148 replied:
I now loop through all server descriptions to get the lowest available wire version.

jyemin commented on driver-sync/src/test/functional/com/mongodb/client/TransactionFailureTest.java:
```diff
@@ -0,0 +1,77 @@
+        ClientSession clientSession = mongoClient.startSession();
+        clientSession.startTransaction();
+        collection.insertOne(clientSession, Document.parse("{_id: 1, a: 1}"));
+        collection.insertOne(clientSession, Document.parse("{_id: 2, a: 1}"));
```
It's not clear which line the test expects the exception to be thrown from.  
To make the intent clearer, remove everything past the first insertOne, since that's the line that should throw.

Except for the ClientSession.close call.  For hygiene, close the clientSession in a finally block.

jstewart148 replied:
I have put the test in a try-finally block.

jyemin reviewed (changes requested):
Reviewed the sync implementation first.  Once we agree on what that looks like, I'll check async.

rozza committed:
Android testing updates

Update evergreen to store the download logs
Set the android license hash

JAVA-3146

jstewart148 committed:
JAVA-2928: code review changes

jstewart148 committed:
JAVA-2928: detect transaction support in CommandMessage

jstewart148 committed:
JAVA-2928: checkstyle fix

jstewart-mongo commented:
I reverted the change for the async changes in OperationExecutorImpl and added the transaction check to CommandMessage.

jyemin commented on driver-core/src/main/com/mongodb/internal/connection/CommandMessage.java:
```diff
@@ -242,6 +245,7 @@ private BsonDocument getCommandToEncode() {
         }
         boolean firstMessageInTransaction = sessionContext.notifyMessageSent();
         if (sessionContext.hasActiveTransaction()) {
+            checkServerVersionForTransactionSupport();
```
Add a unit test in com.mongodb.internal.connection.CommandMessageSpecification.  Probably similar to 'should throw if payload document bigger than max document size'.

jstewart148 replied:
I have added two unit tests in com.mongodb.internal.connection.CommandMessageSpecification, one for each condition where a MongoClientException should be thrown.

jyemin commented on driver-sync/src/main/com/mongodb/client/internal/ClientSessionImpl.java:
```diff
@@ -81,6 +87,8 @@ public void startTransaction() {
 
     @Override
     public void startTransaction(final TransactionOptions transactionOptions) {
+        checkServerVersionForTransactionSupport();
```
Is there any reason to leave this here still?  Does the CommandMessage change cover us?  The behavior is a bit different since the exception will be thrown on first use, but I think that's ok.

jstewart148 replied:
I reverted ClientSessionImpl#startTransaction and the tests passed. `./gradlew check` also passed locally.

jstewart148 replied:
Starting an Evergreen patch: https://evergreen.mongodb.com/version/5c4911500ae6065de57e95d8

jstewart148 committed:
JAVA-2928: revert ClientSessionImpl

rozza commented on driver-async/src/test/functional/com/mongodb/async/client/TransactionFailureTest.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    @BeforeClass
+    public static void beforeClass() {
```
nit: I'd be tempted to remove these noop methods.

jstewart148 replied:
Done

rozza commented on driver-async/src/test/functional/com/mongodb/async/client/TransactionFailureTest.java:
```diff
@@ -0,0 +1,105 @@
+import static com.mongodb.ClusterFixture.serverVersionLessThan;
+import static org.junit.Assume.assumeTrue;
+
+public class TransactionFailureTest {
```
Extend the `DatabaseTestCase` class that will create the MongoClient, MongoDatabase and MongoCollection for you, DRY'ing up some code.

jstewart148 replied:
Done

rozza commented on driver-async/src/test/functional/com/mongodb/async/client/TransactionFailureTest.java:
```diff
@@ -0,0 +1,105 @@
+
+            FutureResultCallback<Void> futureResultCallback = new FutureResultCallback<Void>();
+            collection.insertOne(clientSession, Document.parse("{_id: 1, a: 1}"), futureResultCallback);
+            futureResult(futureResultCallback);
```
Do you need the futureResult method, could you just call `futureResultCallback.get()`?

jstewart148 replied:
I removed the method and check for exceptions in the test method.

rozza commented on driver-sync/src/test/functional/com/mongodb/client/TransactionFailureTest.java:
```diff
@@ -0,0 +1,78 @@
+import static com.mongodb.ClusterFixture.serverVersionLessThan;
+import static org.junit.Assume.assumeTrue;
+
+public class TransactionFailureTest {
```
Same as async test - extend the DatabaseTestCase to DRY up some code.

jstewart148 replied:
Done

rozza reviewed (changes requested):
Looking good - some minor test nits.

jstewart148 committed:
JAVA-2928: test changes

jstewart148 committed:
JAVA-2928: added unit tests

jyemin commented on .evergreen/.evg.yml:
```diff
@@ -121,6 +121,26 @@ functions:
         permissions: public-read
         content_type: ${content_type|application/x-gzip}
 
+  "upload android logs":
```
I think this got in here because you rebased after opening the PR.  Generally good practice not to do that.

jstewart148 replied:
I thought I was in another branch. My mistake.

jyemin commented on driver-async/src/test/functional/com/mongodb/async/client/TransactionFailureTest.java:
```diff
@@ -0,0 +1,74 @@
+            FutureResultCallback<Void> futureResultCallback = new FutureResultCallback<Void>();
+            collection.insertOne(clientSession, Document.parse("{_id: 1, a: 1}"), futureResultCallback);
+            futureResultCallback.get();
+        } catch (MongoClientException e) {
```
I don;t think you need either catch block.  You can just remove the first one, and replace the second with a `throws InterruptedException` on the method definition

jstewart148 replied:
Done

jyemin commented on driver-async/src/test/functional/com/mongodb/async/client/TransactionFailureTest.java:
```diff
@@ -0,0 +1,74 @@
+
+
+    private boolean canRunTests() {
+        return serverVersionLessThan("3.7.0")
```
This should be `serverVersionLessThan("4.0")`

jstewart148 replied:
Done

jyemin commented on driver-core/src/main/com/mongodb/internal/connection/CommandMessage.java:
```diff
@@ -259,6 +263,15 @@ private BsonDocument getCommandToEncode() {
+    private void checkServerVersionForTransactionSupport() {
+        int wireVersion = getSettings().getMaxWireVersion();
+        if (wireVersion < FOUR_DOT_ZERO_WIRE_VERSION
+                || (wireVersion < MAX_DRIVER_WIRE_VERSION && getSettings().getServerType() == SHARD_ROUTER)) {
```
Replace MAX_DRIVER_WIRE_VERSION with FOUR_DOT_TWO_WIRE_VERSION.  Otherwise this will break the next time we have to increment MAX_DRIVER_WIRE_VERSION

jstewart148 replied:
I have added FOUR_DOT_TWO_WIRE_VERSION to com.mongodb.internal.operation.ServerVersionHelper and replaced MAX_DRIVER_WIRE_VERSION with it.

jyemin commented on driver-sync/src/test/functional/com/mongodb/client/TransactionFailureTest.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    private boolean canRunTests() {
+        return serverVersionLessThan("3.7.0")
```
Same comment as on the async tests

jstewart148 replied:
Done

jstewart148 committed:
JAVA-2928: test changes
